pkg/builtin/builtin.go

```diff
@@ -251,13 +265,48 @@ func SysWrite(ctx context.Context, env []string, input string) (string, error) {
 		return "", err
 	}
 
+	// Lock the file to prevent concurrent writes from other tool calls.
+	locker.Lock(params.Filename)
+	defer locker.Unlock(params.Filename)
+
 	data := []byte(params.Content)
 	msg := fmt.Sprintf("Wrote %d bytes to file %s", len(data), params.Filename)
 	log.Debugf(msg)
 
 	return "", os.WriteFile(params.Filename, data, 0644)
 }
 
+func SysAppend(ctx context.Context, env []string, input string) (string, error) {
+	var params struct {
+		Filename string `json:"filename,omitempty"`
+		Content  string `json:"content,omitempty"`
+	}
+	if err := json.Unmarshal([]byte(input), &params); err != nil {
+		return "", err
+	}
+
+	// Lock the file to prevent concurrent writes from other tool calls.
+	locker.Lock(params.Filename)
```

Comment:
Added this for some extra protection against interleaving with concurrent writes, but not sure if it's 100% necessary. I'll test without the lock and drop it if it's not necessary.